drivers/ethernet/Kconfig.xlnx_gem

```diff
@@ -0,0 +1,1387 @@
+#
+# Xilinx Processor System Gigabit Ethernet controller (GEM) driver 
+# configuration options
+#
+# Copyright (c) 2021, Weidmueller Interface GmbH & Co. KG
+# SPDX-License-Identifier: Apache-2.0
+#
+
+DT_COMPAT_XLNX_GEM := xlnx,gem
+
+menuconfig ETH_XLNX_GEM
+	bool "Xilinx GEM Ethernet driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_XLNX_GEM))
+	depends on SOC_XILINX_ZYNQMP_RPU \
+	|| SOC_XILINX_ZYNQMP_APU
+	help
+	  Enable Xilinx GEM Ethernet driver.
+
+config ETH_NIC_MODEL
+	string
+	default "cadence_gem"
+	depends on ETH_XLNX_GEM
+	help
+	  Value of the -nic parameter to be used when starting QEMU.
+	  This parameter specifies which type of Ethernet controller
+	  shall be simulated by QEMU.
+
+# ---------- GEM0 ----------
+
+config ETH_XLNX_GEM_PORT_0
```

Comment:
There are lot of "same" options for the supported ports. Could we move all these settings to DTS in order to avoid this duplication?

Reply:
Hi Jukka, thanks for the quick review, just finished round one of refactoring based on the results of the automated code analysis. From my point of view, all the properties of the controller that can currently be tweaked using menuconfig should work just as well when supplied via the device tree. I just wasn't sure to what degree highly customized bindings are welcome, and if there are any guidelines for identifiers. E.g., adding the well known property "clock-frequency" to one particular device of a class that doesn't have this property by default which requires it probably isn't such a big deal. But for all those little things that can be tweaked, I'd be concerned about re-usability / common usefulness of the corresponding property names. If this is not an issue, I'd give moving the entire configuration to the DT a try. I'm not entirely sure but I doubt that, while improved readability of the code is surely given, this change would eventually eliminate the need to have separate mentions of the maximum of 4 available device instances (in the UltraScale, the Zynq only has GEM0/1) both in the driver's header and the source file. I've tried this before and at some point ran into the situation that I would have needed recursive macros.

Reply:
I've thought about this issue again and one thing came to my mind: there's at least one dependency that is automatically implemented when using Kconfig that I'm not sure of whether it can be expressed when configuring features via the DT or not. The controller supports hardware offloading of checksum generation while transmitting and checksum verification while receiving. Using this feature is generally desireable, so it's set to default y for both directions. The thing is: the QEMU simulation of the GEM, which does have a few quirks anyways, just doesn't implement this. Enable the offloading and you'll see ARP packets going in and out of QEMU, but everything that's checksummed is not going anywhere. Therefore, the checksum offloading items currently have a dependency in Kconfig that is "depends on !QEMU_TARGET". If such a dependency can be expressed in the device tree data, then it should be possible to move the configuration from Kconfig to DT.